feat: document astro features (#3924)
withastro/docs: src/content/docs/en/reference/adapter-reference.mdx

	name: string;
	serverEntrypoint?: string;
	exports?: string[];
  supportedAstroFeatures: AstroFeatureMap;
}

export type SupportsKind = 'unsupported' | 'stable' | 'experimental' | 'deprecated';

export type AstroFeatureMap = {
	/**
	 * The adapter is able to serve static pages
	 */
	staticOutput?: SupportsKind;
	/**
	 * The adapter is able to serve pages that are static or rendered via server
	 */
	hybridOutput?: SupportsKind;
	/**
	 * The adapter is able to serve SSR pages
	 */
	serverOutput?: SupportsKind;
	/**
	 * The adapter can emit static assets
	 */
	assets?: AstroAssetsFeature;
};

export interface AstroAssetsFeature {
	supportKind?: SupportsKind;
	/**

> 	 * Whether if this adapter deploys files in an environment that is compatible with the library `sharp`
> 	 */
> 	isSharpCompatible?: boolean;
> 	/**
> 	 * Whether if this adapter deploys files in an environment that is compatible with the library `squoosh`

Reviewer: ```suggestion
	 * Whether or not this adapter deploys files in an environment that is compatible with the library `sharp`
	 */
	isSharpCompatible?: boolean;
	/**
	 * Whether or not this adapter deploys files in an environment that is compatible with the library `squoosh`
```

* __name__: A unique name for your adapter, used for logging.
* __serverEntrypoint__: The entrypoint for server-side rendering.
* __exports__: An array of named exports when used in conjunction with `createExports` (explained below).

> * __supportedAstroFeatures__: A map of Astro built-in features. Sometime adapters can't support those features, and maybe they want to deprecate them.

Reviewer: ```suggestion
* __supportedAstroFeatures__: A map of Astro built-in features. This allows Astro to determine which features an adapter is unable or unwilling to support so appropriate error messages can be provided.
```
I'm not sure if this is true, but it sounds good!  :sweat_smile: Maybe something like this?

Author: That's a very good suggestion! Thank you!

```

Once you [publish your adapter to npm](https://docs.npmjs.com/cli/v8/commands/npm-publish), running `astro add example` will install your package with any peer dependencies specified in your `package.json`. We will also instruct users to update their project config manually.

## Astro features

<Since v="3.0.0" />

> Astro features are a way for an adapter to tell Astro if they are able to support a feature - and its level of support.  

Reviewer: ```suggestion
Astro features are a way for an adapter to tell Astro whether they are able to support a feature, and also the adapter's level of support.  
```

When using these properties, Astro will:
- run specific validation; 
- emit contextual to the logs;

These operations are run based on the features supported or not supported, their level of support, and the configuration that the user uses.

> ```js title="my-adapter.mjs" ins={9-15}
> export default function createIntegration() {
>   return {
>     name: '@matthewp/my-adapter',
>     hooks: {
>       'astro:config:done': ({ setAdapter }) => {
>         setAdapter({
>           name: '@matthewp/my-adapter',
>           serverEntrypoint: '@matthewp/my-adapter/server.js',
>           supportedAstroFeatures: {
>             assets: {
>               supportKind: "experimental",
> 	            isSharpCompatible: false,
> 	            isSquooshCompatible: false
>             }
>           }  
>         });
>       },
>     },
>   };
> }
> ```
> ```
> 
> In this example, we're telling Astro that this adapter has experimental support for assets, but it's not compatible with the built-in services.

Reviewer: ``````suggestion
The following configuration tells Astro that this adapter has experimental support for assets, but the adapter is not compatible with the built-in services Sharp and Squoosh:

```js title="my-adapter.mjs" ins={9-15}
export default function createIntegration() {
  return {
    name: '@matthewp/my-adapter',
    hooks: {
      'astro:config:done': ({ setAdapter }) => {
        setAdapter({
          name: '@matthewp/my-adapter',
          serverEntrypoint: '@matthewp/my-adapter/server.js',
          supportedAstroFeatures: {
            assets: {
              supportKind: "experimental",
	            isSharpCompatible: false,
	            isSquooshCompatible: false
            }
          }  
        });
      },
    },
  };
}
```
``````
I just changed the order around, so people know what to look for in the code sample first!

> Astro will log a **warning** to the terminal

Reviewer: ```suggestion
Astro will log a **warning** to the terminal:
```

```
[@matthewp/my-adapter] The feature is experimental and subject to issues or changes.
```

> and an error if the service used for assets is not compatible with the adapter

Reviewer: ```suggestion
and an error if the service used for assets is not compatible with the adapter:
```
This gets around having a sentence broken up.  :joy:
